src/sync/impl/apple/system_configuration.hpp

```diff
@@ -0,0 +1,70 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2017 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_OS_SYSTEM_CONFIGURATION_HPP
+#define REALM_OS_SYSTEM_CONFIGURATION_HPP
+
+#include "sync/impl/network_reachability.hpp"
+
+#if NETWORK_REACHABILITY_AVAILABLE
+
+#include <SystemConfiguration/SystemConfiguration.h>
+
+namespace realm {
+
+class SystemConfiguration {
+public:
+    static SystemConfiguration& shared();
+
+    SCNetworkReachabilityRef network_reachability_create_with_name(CFAllocatorRef allocator,
+                                                                   const char *hostname);
+    SCNetworkReachabilityRef network_reachability_create_with_address(CFAllocatorRef allocator,
+                                                                      const sockaddr *address);
+    bool network_reachability_set_dispatch_queue(SCNetworkReachabilityRef target,
+                                                 dispatch_queue_t queue);
+    bool network_reachability_set_callback(SCNetworkReachabilityRef target,
+                                           SCNetworkReachabilityCallBack callback,
+                                           SCNetworkReachabilityContext *context);
+    bool network_reachability_get_flags(SCNetworkReachabilityRef target,
+                                        SCNetworkReachabilityFlags *flags);
+
+private:
+    using network_reachability_create_with_name_t = SCNetworkReachabilityRef(*)(CFAllocatorRef, const char*);
```

Comment:
I wonder if we can cheat here and use `decltype(SCNetworkReachabilityCreateWithName)` as the RHS rather than having to restate the type?

I'd be inclined to drop the `network_reachability` prefix from the private members here to reduce the verbosity.